### PR TITLE
Use Auth0 instead of HTTP Basic for authentication

### PR DESCRIPTION
**Pull request by mozbhearsum**

### CODE REVIEW

**Comment by mozbhearsum** [comment]
There's still a bit of work to be done here, but in the interest of expediency, could you take a look at the current state of things, @nthomas-mozilla? Here's the high level summary:
* The authentication flow is done fully on the frontend, ultimately resulting in an access_token (aka Bearer token) that is sent to the backend for any post/put/delete requests.
* The backend verifies it cryptographically, using the keys that auth0 publishes in its `.well_known` directory. It also uses auth0's `userinfo` endpoint to pull additional profile information (email address + picture is what we currently use from there).
* Switch to 8010 as our local development port, because that's what the auth0 local dev account requires.
* Add a mechanism for adding a named local admin user (`docker-compose run balrogadmin create-local-admin`). This is now required, because local dev goes through auth0, and we cannot predict the name of the admin at image creation time.
* Most of the UI stuff is cribbed from the auth0 quickstart: https://auth0.com/docs/quickstart/spa/angularjs/05-token-renewal
* There's still a bit of work to be done for the frontend auth0 config -- we decided with CloudOps that they would be providing the clientId and redirectUri in the environment, and possibly the other values as well (as opposed to the configs that are currently baked into the repo).

I know the Auth stuff is really weird and complex, so let me know if there's any confusion about it.

**Comment by mozbhearsum** [comment]
@nthomas-mozilla, can you have a look at this when you have time? We're pretty sure this is in its final state now. We've done as much testing as possible locally & on an EC2 instance, and everything appears to be working. This patch includes Auth0 support for the backend, frontend, agent, and client tools. The client will initially support sending HTTP Basic and a Bearer token, to avoid being in a state where the client and backend disagree on how auth needs to be done at any point in time. The rollout plan is as follows:

1. Ask CloudOps to add new environment variables required for Auth0 configuration. I think it's best for this to be done in its own deployment, so they exist before the new code lands.
2. Roll a new "balrogclient" package plus new "balrogscript" & deploy that. Those workers will then begin to send HTTP Basic and a Bearer token, with the latter being ignored.
3. Deploy new backend, frontend, and agent. We will also ask CloudOps to disable HTTP Basic at this time.
4. Update the clients and agent to stop sending HTTP Basic, as it will not even be noticed by Balrog anymore.

We ran through these steps on the EC2 instance and they worked well. We'll do the same with balrog dev and stage when this patch is merged before going to prod, as well as run a staging release so more fully exercise the client code paths. If you'd like to poke at the EC2 instance it's available at https://10.134.44.85:8010/ (see the note below).

One caveat I need to mention is that the Balrog Agent is currently broken for local development due to the fact that it relies on a machine-to-machine clientId/secret, and even though it's locked to localhost, we've been told that we shouldn't publish that secret, just in case of config issues or bugs on the Auth0 side in the future. I don't know how we're going to fix this yet, but I _think_ we can live with this for a week or two, although I'd like to fix it faster. If you think this should block merging, please feel free to push back on this.

On the off-chance you hit it, I should also mention that there was one strange issue we hit on the EC2 instance, but I strongly suspect that is caused by the Auth0 trial account I was using it with, and not an issue with the code (it doesn't happen in local testing, which uses the official Mozilla Auth0 account). I won't be 100% sure of this until we deploy to dev, unfortunately.

File: auslib/web/admin/views/permissions.py [hunk 1]
```diff
@@ -24,29 +26,23 @@ def get(self):
...
         # to retrieve permissions of anyone, it may make privilege escalation
         # attacks easier.
         # TODO: do this at the database layer
-        else:
-            if username != current_user and not dbo.hasPermission(current_user, "permission", "view"):
-                return problem(status=403, title="Forbidden",
-                               detail="You are not authorized to view permissions of other users.")
+        if username != current_user and not dbo.hasPermission(current_user, "permission", "view"):
+            return problem(status=403, title="Forbidden",
+                           detail="You are not authorized to view permissions of other users.")
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply auslib/web/admin/views/permissions.py]: I honestly don't. I'm willing to bet it was an oversight, though.

File: auslib/util/auth.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+def get_jwks(auth_domain):
+    jwks_url = "https://{}/.well-known/jwks.json".format(auth_domain)
+    req = requests.get(jwks_url)
+    req.raise_for_status()
+    return req.json()
+
+
+@lru_cache(2048)
+def get_additional_userinfo(auth_domain, access_token):
+    return json.loads(auth0_Users(auth_domain).userinfo(access_token))
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply auslib/util/auth.py]: AFAICT, the access_token in the args will be stored in the cache, so this is worth considering. I'm trying to find out how long the access tokens stay valid for to help decide how much we should worry about this.

File: ui/app/js/services/auth0_service.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          }
+        } else {
+          localLogin(result);
+        }
+      }
+    );
+  }
+  function scheduleRenewal() {
+    // Renew tokens 5 minutes ahead of expiration
+    var delay = localStorage.getItem("expiresAt") - Date.now() - 500000;
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply ui/app/js/services/auth0_service.js]: :man_facepalming:

File: docker-compose-py3.yml [hunk 4]
```diff
@@ -78,12 +86,19 @@ services:
...
       - BALROG_PASSWORD=na
       - TELEMETRY_API_ROOT=abc
       - LOG_FORMAT=plain
       - LOG_LEVEL=WARNING
+      - AUTH0_DOMAIN=auth.mozilla.auth0.com
+      - ALLOW_COOKIE_FROM_IP_URL=1
+      - AUTH0_AUDIENCE=balrog-localdev
+      - AUTH0_M2M_CLIENT_ID=xWFk4cJVfLm3Vg7tFIK9H8j6LeFmsF3B
+      # If you don't have this, ask.
+      - AUTH0_M2M_CLIENT_SECRET
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply docker-compose-py3.yml]: Will do.

File: ui/app/js/config.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// This file is not packaged with the app; it must be provided as part of the deployment environment
+angular.module("config", [])
+.constant("Auth0Config", {
+    domain: "auth.mozilla.auth0.com",
+    audience: "login.taskcluster.net",
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply ui/app/js/config.js]: Placeholder. In fact, this file shouldn't be checked in at all (we generated when the admin app starts up, as you probably saw).

File: ui/config/env/prod.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+    "Auth0Config": {
+        "clientID": "TODO",
+        "redirectUri": "TODO"
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply ui/config/env/prod.json]: Sorry, this was supposed to have been removed as well. I thought I purged all of this first attempt at environment configs...

**Comment by mozbhearsum** [comment]
> A question - is it possible to manually-test scheduled changes needing multiple signoffs ? Perhaps that's possible in local-dev if you might use two separate accounts in your control.

Yeah, that's how'd you have to do it. I think this is actually an improvement over the pre-nginx state of things, where you'd have to change the hardcoded REMOTE_USER and restart the containers to change usernames. Of course, it does mean you need multiple accounts, and for the usernames to differ.

File: client/balrogclient/api.py [hunk 7]
```diff
@@ -122,9 +170,14 @@ def do_request(self, url, data, method):
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'}
         before = time.time()
+        if self.auth0_secrets:
+            access_token = _get_auth0_token(self.auth0_secrets)
```
[the remark this answers is not in the document]
Comment (rail) [reply client/balrogclient/api.py]: I'll add caching and checking the expiration.

File: auslib/util/auth.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            "description": "Authorization header must be Bearer token"},
+            401)
+
+    token = parts[1]
+    return token
+
+
+@lru_cache(2048)
+def get_jwks(auth_domain):
+    jwks_url = "https://{}/.well-known/jwks.json".format(auth_domain)
```
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply auslib/util/auth.py]: Yeah, this is going to be a one item cache. Here's some additional info on the keys (from https://auth0.com/docs/jwks):

`
Currently Auth0 only supports a single JWK for signing, however, it is important to assume this endpoint could contain multiple JWKs. As an example, multiple keys can be found in the JWKS when rotating signing certificates.
`

So, a timeout might be a good precaution, maybe a 1h timeout or so? It's a pretty cheap download, so lowering it further probably wouldn't hurt.

File: auslib/util/auth.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mozbhearsum) [reply auslib/util/auth.py]: They'll be valid for a max of 15min in prod, so this is probably OK?

**Comment by mozbhearsum** [comment]
I think we've addressed all of the feedback now...

**Comment by mozbhearsum** [comment]
Sorry, had to push one last patch (really!) after @gdestuynder told me we need to keep our in-app CSRF protection, because the auth0 version only covers the login process, not the overall app.

File: auslib/util/auth.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nthomas-mozilla) [reply auslib/util/auth.py]: That seems fine to me. It's probably uncommon and there's a unique error message when we've cached stale data.

File: auslib/util/auth.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nthomas-mozilla) [reply auslib/util/auth.py]: Seems like a small window. Could always confirm with some of the sec folks.

**Comment by nthomas-mozilla** [comment]
Oh, one thing as a followup.  Initial setup of a new database will need the `create-local-admin` trick to bootstrap, right ? Would be good to update withoutdocker.rst for that.

**Pull request merged**